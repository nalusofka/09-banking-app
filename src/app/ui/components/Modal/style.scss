@use '../../styles/variables' as var;

.modal-backdrop {
    align-items: center;
    background: rgba(0, 0, 0, 0.5);
    display: flex;
    height: 100%;
    justify-content: center;
    left: 0;
    position: fixed;
    top: 0;
    width: 100%;
    z-index: 999;
  }
  
  .modal {
    background: var.$color-white;
    border-radius: 8px;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.2);
    max-width: 400px;
    padding: 20px;
    width: 100%;
  }
  
  .modal-title {
    font-size: 1.5rem;
    margin-bottom: 10px;
  }
  
  .modal-content {
    margin-bottom: 20px;
  }
  
  .modal-actions {
    display: flex;
    gap: 10px;
    justify-content: flex-end;
  }
  
  .modal-button {
    background: var.$color-white;
    border-radius: 4px;
    border: none;
    color: var.$color-dark-blue;
    cursor: pointer;
    padding: 10px 20px;
  }
