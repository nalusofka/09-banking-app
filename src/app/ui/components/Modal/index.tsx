import React from "react";
import "./style.scss";
import Title from "../Title";
import Body from "../Body";
import Button from "../Button";

interface ModalProps {
  isOpen: boolean;
  onClose: () => void;
  title: string;
  children: React.ReactNode;
  onConfirm?: () => void;
}

const Modal: React.FC<ModalProps> = ({ isOpen, onClose, title, children, onConfirm }) => {
  if (!isOpen) return null;

  return (
    <div className="modal-backdrop">
      <div className="modal">
        <Title as="h4" color="primary">{title}</Title>
        <br />
        <Body color="primary">{children}</Body>
        <div className="modal-actions">
          <Button className="modal-button" onClick={onClose}>
            Cancelar
          </Button>
          {onConfirm && (
            <Button onClick={onConfirm}>Confirmar</Button>
          )}
        </div>
      </div>
    </div>
  );
};

export default Modal;
