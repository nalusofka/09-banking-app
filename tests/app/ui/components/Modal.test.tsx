import React from "react";
import "@testing-library/jest-dom";
import { render, fireEvent } from "@testing-library/react";
import Modal from "../../../../src/app/ui/components/Modal";

describe("Componente Modal", () => {
    const mockOnClose = vi.fn();
    const mockOnConfirm = vi.fn();

    afterEach(() => {
        vi.clearAllMocks();
    });

    it("no se renderiza cuando `isOpen` es `false`", () => {
        const { container } = render(
            <Modal isOpen={false} onClose={mockOnClose} title="Título del Modal">
                Contenido del Modal
            </Modal>
        );

        expect(container.firstChild).toBeNull();
    });

    it("se renderiza correctamente cuando `isOpen` es `true`", () => {
        const { getByText, asFragment } = render(
            <Modal isOpen={true} onClose={mockOnClose} title="Título del Modal">
                Contenido del Modal
            </Modal>
        );

        expect(getByText("Título del Modal")).toBeInTheDocument();
        expect(getByText("Contenido del Modal")).toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot();
    });

    it("ejecuta `onClose` al hacer clic en el botón Cancelar", () => {
        const { getByText } = render(
            <Modal isOpen={true} onClose={mockOnClose} title="Título del Modal">
                Contenido del Modal
            </Modal>
        );

        const cancelButton = getByText("Cancelar");
        fireEvent.click(cancelButton);

        expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it("renderiza y ejecuta `onConfirm` si se proporciona", () => {
        const { getByText } = render(
            <Modal
                isOpen={true}
                onClose={mockOnClose}
                onConfirm={mockOnConfirm}
                title="Título del Modal"
            >
                Contenido del Modal
            </Modal>
        );

        const confirmButton = getByText("Confirmar");
        fireEvent.click(confirmButton);

        expect(mockOnConfirm).toHaveBeenCalledTimes(1);
    });

    it("no renderiza el botón Confirmar si `onConfirm` no se proporciona", () => {
        const { queryByText } = render(
            <Modal isOpen={true} onClose={mockOnClose} title="Título del Modal">
                Contenido del Modal
            </Modal>
        );

        const confirmButton = queryByText("Confirmar");
        expect(confirmButton).not.toBeInTheDocument();
    });
});
