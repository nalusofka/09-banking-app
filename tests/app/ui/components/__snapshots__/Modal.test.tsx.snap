// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente Modal > se renderiza correctamente cuando \`isOpen\` es \`true\` 1`] = `
<DocumentFragment>
  <div
    class="modal-backdrop"
  >
    <div
      class="modal"
    >
      <h4
        class="title title-primary"
        data-testid="title-component"
      >
        TÃ­tulo del Modal
      </h4>
      <br />
      <p
        class="body' body-primary"
        data-testid="body-component"
      >
        Contenido del Modal
      </p>
      <div
        class="modal-actions"
      >
        <button
          aria-disabled="false"
          class="modal-button"
          data-testid="button-component"
          type="button"
        >
          Cancelar
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
